src/v/cloud_storage/materialized_segments.cc

```diff
@@ -238,7 +178,13 @@ void materialized_segments::trim_readers(size_t target_free) {
         // Readers hold a reference to the segment, so for the
         // segment.owned() check to pass, we need to clear them out.
         while (!st.readers.empty() && _reader_units.current() < target_free) {
-            evict_reader(std::move(st.readers.front()));
+            auto partition = st.parent;
+            // TODO: consider asserting here instead: it's a bug for
+            // 'partition' to be null, since readers outlive the partition, and
```

Comment:
Do you mean readers _cannot_ outlive the partition?

Reply:
Oops, yes. Done